winrm/protocol.py

```diff
@@ -395,6 +395,31 @@ def cleanup_command(self, shell_id, command_id):
         # TODO change assert into user-friendly exception
         assert uuid.UUID(relates_to.replace('uuid:', '')) == message_id
 
+    def send_command_input(self, shell_id, command_id, stdin_input):
```

Comment:
Setting END=True would mean that there is no more data coming in on stdin.  However, we cannot make the determination that the user is done sending input.  Maybe they want to send input, get output, then send more input based on the output. Closing the stream by setting END=True would prevent additional input from being sent to the running process would it not?

Reply:
That's why it's a kwarg, it's set to True by default but if the user needs to send more input then they set it to `False`. Without this attribute I cannot run a command like `powershell.exe -` as it's forever waiting for the stdin to finish sending it's input.

Reply:
Testing current code as is, by sending multiple different instances of input using cmd
```
p.writeln("echo TEST1")
p.writeln("echo TEST2")
p.writeln("echo TEST3")
p.writeln("echo TEST4")
p.writeln("echo hello world && exit")
print(p.get_stdout())
h.disconnect()

output:
Microsoft Windows [Version 10.0.17763.107]
(c) 2018 Microsoft Corporation. All rights reserved.
C:\Users\Administrator>echo TEST1
TEST1
C:\Users\Administrator>echo TEST2
TEST2
C:\Users\Administrator>echo TEST3
TEST3
C:\Users\Administrator>echo TEST4
TEST4
C:\Users\Administrator>echo hello world && exit
hello world 
```

Conducting the same test using powershell:
```
p.writeln("echo TEST1")
p.writeln("echo TEST2")
p.writeln("echo TEST3")
p.writeln("echo TEST4")
p.writeln('echo "hello world"')
p.writeln("$PSVersionTable.PSVersion")
p.writeln('exit')
print(p.get_stdout())

output:
Windows PowerShell 
Copyright (C) Microsoft Corporation. All rights reserved.
echo TEST1
TEST1
PS C:\Users\Administrator> echo TEST2
TEST2
PS C:\Users\Administrator> echo TEST3
TEST3
PS C:\Users\Administrator> echo TEST4
TEST4
PS C:\Users\Administrator> echo "hello world"
hello world
PS C:\Users\Administrator> $PSVersionTable.PSVersion
Major  Minor  Build  Revision
-----  -----  -----  --------
5      1      17763  1       
PS C:\Users\Administrator> 
```
If you open a process which is going to expect input before exiting, your process shouldn't terminate until you provide input.

Reply:
So what's the executable and arguments you are running? I get your use case where you need to send input, display output and repeat so you would set `End="false"` but when I'm testing a similar script locally, if I omit the `End` attribute altogether the process never finishes even after sending my input so it hangs waiting for a response.

Maybe instead of defaulting to `True` we can default it to `False` but in my testing it seems to be required.

Reply:
Essentially:
```
protocol.run_command(self.shell_id, "powershell")
```

Reply:
Looking for the kwarg, I can't seem to find the kwarg anywhere in send_message for end=True.

Reply:
Question: are you using ```get_command_output```? If you use ```get_command_output```, this function will wait till the command has exited before giving you the full output of the command.
```_raw_get_command_output``` will get you the output immediately without blocking.

Reply:
> Looking for the kwarg, I can't seem to find the kwarg anywhere in send_message for end=True.

I'm suggesting you add the kwarg to `send_command_input` as `End` is an attribute on the `Send` body as per https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/connection/winrm.py#L432.

> are you using get_command_output?

I'm trying to run just that and it hangs on the `get_command_output` as you said the process won't exit because it's still waiting on input. If you close the input pipe with the `End` attribute then the process will finish. So this shows you that there is a use case for End being either `True` or `False`.

Reply:
Well, ```get_command_output``` will wait for the process to finish.  If you are running a windows executable like, ```cmd``` or ```powershell```, unless you send ```exit``` to stdin, those programs will run forever.  Therefore, if you run, ```get_command_output``` without having sent proper input to the running program to exit, it is suppose to run forever.  Essentially, what you have done is open up a command prompt window, typed something in, and then started waiting for the window to close.  Of course, however, the program will never exit and the window will never close until someone types "exit" into the input.  Essentially what you have done by closing the stdin pipe line, which underneath the hood is a named pipe to windows, is closed the named pipe which then closes the running powershell and/or cmd process because those processes are smart enough to know what to do when their named pipe for stdin is closed.  However, this is entirely dependent on the executable that you are running and whether or not the developer of the executable has code added that will also close their application by closing the stdin named pipe.
Can you try your test by running the ```_raw_get_command_output```?
You may have to try and except for the timeout, but you should be able to interact with the process interactively without waiting for ```get_command_output``` because you never use it.

Telling a process that it should exit because you closed it's stdin named pipe doesn't sound like a reliable means of doing anything besides actually closing the named pipe which isn't a common use case and not the one I'm targeting here; maybe a future feature if needed perhaps.  Then again, it's windows and they do all sorts of wacky (INSERT SWEAR WORD OF CHOICE HERE).

Reply:
You are still thinking of an interactive process which I agree we don't want to close the stdin pipe but there are still valid use cases for closing the pipe which is why this kwarg should be added. To prove this try and run the following powershell script

```
begin {
    $input_list = [System.Collections.Generic.List[String]]@()
}
process {
    $input_list.Add($input)
}
end {
    $input_list -join ", "
}
```

Without closing the stdin pipe it is impossible for this script to finish as it is constantly waiting for input and I can prove that to you.

```
import base64

from winrm import Protocol


script_with_input = base64.b64encode(u"""begin {
    $input_list = [System.Collections.Generic.List[String]]@()
}
process {
    $input_list.Add($input)
}
end {
    $input_list -join ", "
}
""".encode('utf-16-le')).decode('utf-8')

p = Protocol("http://server:5985/wsman", username="username", password="password")
shell_id = p.open_shell()
cmd_id = p.run_command(shell_id, "powershell", ["-EncodedCommand", script_with_input])
for i in range(0, 5):
    p.send_command_input(shell_id, cmd_id, str(i) + "\r\n", end=(i == 4))

out = p.get_command_output(shell_id, cmd_id)
p.cleanup_command(shell_id, cmd_id)
p.close_shell(shell_id)

print(out)
```

I understand why you don't have it and it may not be required but I'm telling you why it can be needed for some scenarios. As an FYI having this is a key enabler of improving file transfer through pywinrm if someone eventually wants to implement that.

Reply:
> As an FYI having this is a key enabler of improving file transfer through pywinrm if someone eventually wants to implement that.

That may be, and it's a much better argument than the code segment you provided because the code example you provided doesn't use ```_raw_get_command_output```.  Esentially, you have made an encoded command which depends on input being sent, and then you call ```get_command_output``` which is just going to wait forever.  Unless you close the stdin named pipe which, very coincidentally I might add, ends up closing the process; a behavior that is entirely dependent on the developers of the executable you are running. IE, it's a total crap shoot as to what will really happen on most executables.

It also sounds like an additional feature.  Is this something we can add as an additional feature in a different merge request?
That way, after this request is merged, you can totally add it if you see fit.  Especially if it's going to be needed for file transfers.

Reply:
Sorry, just catching up on this PR. Also note, if I misinterpret something please let me know.

Essentially I agree with @jborean93 that a kwarg for `end` should be provided. As I think through this, I am not seeing any drawback of including it as a default of `False`.

I understand the concerns about developers essentially shooting themselves in the foot, but I don't see that as pywinrm's responsibility. Preventing those scenarios should be within the upstream application(and each one has a different use case).

My view of pywinrm ultimately is to help fully utlize pywinrm fully in a Pythonic way, which adding an `end` would provide that. It is allowed in the payload, and I would rather add it right the first time even if we will need to make changes elsewhere later on. Adding the kwarg doesn't hurt anything or even change the behavior is unspecified. I would also home that either through knowledge or testing, they would understand what is going on.

My vote is to add this right the first time, and and address any issues where developers may shoot themselves in the foot with future PRs.

Reply:
I view it as feature creep that can be addressed in a future merge.  However, just as well, if both @badcure and @jborean93 see this as an absolutely necessary feature that should be done at the same time as this, then in it shall go in spite of my personal views on the situation.  It will default to "False", to help developers not shoot their own feet, but the ability to overwrite this value will be available through kwargs.  I'm going to push some changes up right now, but can both @badcure and @jborean93 double check my work to make sure that this what you both want?

Reply:
Sorry for the delay, I got working on other things which got in the way of these changes.  I just got around to pushing them up. @badcure and @jborean93 , Do these changes meet with your satisfaction?  Do you need some time to "kick the wheels" so to speak?

Reply:
Before committing, my testing of sending input to an already closed stream resulted in the following:
```
File "C:\Program Files\Python37\lib\site-packages\pywinrm-0.3.1.dev0-py3.7.egg\winrm\protocol.py", line 281, in send_message
winrm.exceptions.WinRMError: The handle is invalid.  (extended fault data: {'transport_message': 'Bad HTTP response returned from server. Code 500', 'http_status_code': 500, 'wsmanfault_code': '6', 'fault_code': 's:Receiver', 'fault_subcode': 'w:InternalError'})

```
This, to me, confirms that the stream closes when asked as well as returns an appropriate error if attempting to use a closed stream.